### ISSUE TITLE
investigate why rexplorer as it is, doesn't work on testnet

### ISSUE THREAD

**Issue by GlenDC** [threefoldfoundation/rexplorer#1]
It seems to try to revert —as well as spend— non-existing (or already spend?) coin outputs.

**Comment by GlenDC** [comment]
You can reproduce the (first) issue easily yourself as follows:

```
rexplorer --redis-db 1 --network testnet
2018/08/08 11:01:06 starting rexplorer v0.1.0...
2018/08/08 11:01:06 loading rivine gateway module (1/3)...
2018/08/08 11:01:06 loading rivine consensus module (2/3)...
2018/08/08 11:01:06 loading internal explorer module (3/3)...
2018/08/08 11:01:51 [ERROR] either all owners should have the multisig address linked or none, have 2/3
2018/08/08 11:01:53 Closing consensus module...
2018/08/08 11:01:53 Closing gateway module...
panic: failed to revert coin output c03c00245baecf16c1e30a745b49a92aeffefd479e910e87182edc9591197c71: redis: failed to revert coin output: cannot drop coin output c03c00245baecf16c1e30a745b49a92aeffefd479e910e87182edc9591197c71 in tfchain:testnet:ucos: redigo: nil returned
goroutine 1 [running]:
main.(*Explorer).ProcessConsensusChange(0xc420276780, 0xb0f09a983f85ca90, 0xd84701b2bdd63669, 0x84b39cfee546042d, 0x779020875b6e2170, 0xc42077a000, 0x583, 0x5b6, 0xc4209d4000, 0x584, ...)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/explorer.go:123 +0x20d6
github.com/threefoldfoundation/rexplorer/vendor/github.com/rivine/rivine/modules/consensus.(*ConsensusSet).initializeSubscribe.func1(0xc42028e0e0, 0x15137a8, 0xc42028e0e0)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/vendor/github.com/rivine/rivine/modules/consensus/subscribe.go:137 +0x261
github.com/threefoldfoundation/rexplorer/vendor/github.com/rivine/bbolt.(*DB).View(0xc420328000, 0xc420231440, 0x0, 0x0)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/vendor/github.com/rivine/bbolt/db.go:708 +0x90
github.com/threefoldfoundation/rexplorer/vendor/github.com/rivine/rivine/modules/consensus.(*ConsensusSet).initializeSubscribe(0xc420045400, 0x1549f00, 0xc420276780, 0x0, 0x0, 0x0, 0x0, 0x1, 0x280)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/vendor/github.com/rivine/rivine/modules/consensus/subscribe.go:95 +0xa3
github.com/threefoldfoundation/rexplorer/vendor/github.com/rivine/rivine/modules/consensus.(*ConsensusSet).ConsensusSetSubscribe(0xc420045400, 0x1549f00, 0xc420276780, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/vendor/github.com/rivine/rivine/modules/consensus/subscribe.go:161 +0x164
main.NewExplorer(0x1551520, 0xc4200b9b80, 0x1552780, 0xc420045400, 0x14d1808, 0x7, 0x7ffeefbffbb1, 0x7, 0x14ce046, 0x3, ...)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/explorer.go:70 +0x262
main.(*Commands).Root(0xc4202325a0, 0xc420276000, 0xc4200a26c0, 0x0, 0x4, 0x0, 0x0)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/commands.go:80 +0x783
main.(*Commands).Root-fm(0xc420276000, 0xc4200a26c0, 0x0, 0x4, 0x0, 0x0)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/main.go:51 +0x52
github.com/threefoldfoundation/rexplorer/vendor/github.com/spf13/cobra.(*Command).execute(0xc420276000, 0xc4200a8100, 0x4, 0x4, 0xc420276000, 0xc4200a8100)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/vendor/github.com/spf13/cobra/command.go:762 +0x468
github.com/threefoldfoundation/rexplorer/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420276000, 0x1, 0x14ebd5f, 0x4f)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/vendor/github.com/spf13/cobra/command.go:852 +0x30a
github.com/threefoldfoundation/rexplorer/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420276000, 0xc4202325b0, 0x14d13af)
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
        /Users/glendc/go/src/github.com/threefoldfoundation/rexplorer/main.go:99 +0x6c4
```

At that point you'll have the following `tfchain/testnet` stats:

```
tfchain/testnet has:
  * a total of 100558178.90000007 TFT, of which 100549977.50000007 TFT is liquid,
    8201.4 TFT is locked and 558178.90000007 TFT is paid out as fees and miner rewards
  * 99.99184% liquid coins of a total of 100558178.90000007 TFT coins
  * 00.00816% locked coins of a total of 100558178.90000007 TFT coins
  * a block height of 55816, with the time of the highest block being 2018-06-26 16:33:28 +0200 CEST (1530023608)
  * a total of 55817 blocks, 196 value transactions and 1511 coin inputs
  * a total of 56346 coin outputs, of which 55610 are liquid, 736 are locked and 55981 are payouts/fees
  * a total of 275 unique wallet addresses that have been used
  * an average of 01.86224% wallet coin outputs per value transaction
  * an average of 00.00351% value transactions per block
  * 98.69378% liquid outputs of a total of 56346 coin outputs
  * 01.30622% locked outputs of a total of 56346 coin outputs
  * 00.34994% value transactions of a total of 56010 transactions
```

**Comment by GlenDC** [comment]
Adding some debug statements we get to see the following:

```
2018/08/08 11:12:42 [DEBUG] [height:55688] adding coin output "c03c00245baecf16c1e30a745b49a92aeffefd479e910e87182edc9591197c71": {"value":"800000000","condition":{"type":1,"data":{"unlockhash":"0166358d9c0efc3fca196df46e8b985e9fc0696e0b5b10d8d5d84eddeddbcc4b6ad60bf95fcb70"}}}
2018/08/08 11:12:43 [DEBUG] [height:55696] spend coin output "c03c00245baecf16c1e30a745b49a92aeffefd479e910e87182edc9591197c71"
2018/08/08 11:12:43 [DEBUG] [height:55816] revert coin output "c03c00245baecf16c1e30a745b49a92aeffefd479e910e87182edc9591197c71"
```

So our issue is that reverted coin outputs can also have been spend already, so we have to track spend coin outputs as well, and properly revert them in the user's wallet.

**Comment by GlenDC** [comment]
Fixing this, by tracking spent coin outputs, we get following panic:

```
panic: failed to spend coin output 8e1dee8a447e55c39f3f29595336a5617d88c1f31dac1dd064eb7250d3bf92a5: redis: failed to spend coin output: cannot drop coin output 8e1dee8a447e55c39f3f29595336a5617d88c1f31dac1dd064eb7250d3bf92a5 in tfchain:testnet:ucos: redigo: nil returned
```

with the stats at that point being:

```
tfchain/testnet has:
  * a total of 100550011.00000011 TFT, of which 100541811.00000011 TFT is liquid,
    8200 TFT is locked and 559305.60000007 TFT is paid out as fees and miner rewards
  * 99.99184% liquid coins of a total of 100550011.00000011 TFT coins
  * 00.00816% locked coins of a total of 100550011.00000011 TFT coins
  * a block height of 55929, with the time of the highest block being 2018-06-26 16:30:19 +0200 CEST (1530023419)
  * a total of 55930 blocks, 163 value transactions and 1465 coin inputs
  * a total of 56373 coin outputs, of which 55651 are liquid, 722 are locked and 56062 are payouts/fees
  * a total of 275 unique wallet addresses that have been used
  * an average of 01.90798% wallet coin outputs per value transaction
  * an average of 00.00291% value transactions per block
  * 98.71925% liquid outputs of a total of 56373 coin outputs
  * 01.28075% locked outputs of a total of 56373 coin outputs
  * 00.29060% value transactions of a total of 56090 transactions
```

**Comment by GlenDC** [comment]
Adding again some debug statements we get the following info:

```
2018/08/08 12:05:27 [DEBUG] [height:53231] adding coin output "8e1dee8a447e55c39f3f29595336a5617d88c1f31dac1dd064eb7250d3bf92a5": {"value":"999900000000","condition":{"type":1,"data":{"unlockhash":"0114df42a3bb8303a745d23c47062a1333246b3adac446e6d62f4de74f5223faf4c2da465e76af"}}}
2018/08/08 12:05:28 [ERROR] either all owners should have the multisig address linked or none, have 2/3
2018/08/08 12:05:29 [DEBUG] [height:55814] spend coin output "8e1dee8a447e55c39f3f29595336a5617d88c1f31dac1dd064eb7250d3bf92a5"
2018/08/08 12:05:30 [DEBUG] [height:55929] spend coin output "8e1dee8a447e55c39f3f29595336a5617d88c1f31dac1dd064eb7250d3bf92a5"
```

Which seems to indicate a double spend.

**Comment by GlenDC** [comment]
Turns out no double spending, which makes sense, as that would have been a weird one to have slipped through both Rivine and TFchain.

The real issue was that we didn't revert coin inputs (essentially the reverse of SpendCoinOutput), which meant that it looked like a double spend, but what had really happened is that we spent, reverted and spent it for real. It's just that the revert wasn't tracked, hence why it looked like a double spend.

Further there was a small bug causing the auto-unlocking of timelocked outputs most of the time.
